source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc

```diff
@@ -91,9 +103,21 @@ void UpstreamProxyProtocolSocket::generateHeaderV2() {
   if (!options_ || !options_->proxyProtocolOptions().has_value()) {
     Common::ProxyProtocol::generateV2LocalHeader(header_buffer_);
   } else {
+    // Process any custom TLVs from the host metadata.
+    processCustomTLVsFromHost();
+    // Populate any custom TLVs defined the config that are not in the host metadata.
+    for (const auto& tlv : config_tlvs_) {
+      // Skip any TLV that is already in the custom TLVs. We want the host
+      // metadata value to take precedence when there is a conflict.
+      if (host_metadata_tlv_types_.contains(tlv.type)) {
```

Comment:
I think you can move this loop into `processCustomTLVsFromHost` (and rename to `processCustomTLVs`) and then `host_metadata_tlv_types_` can also be a local variable, not a member.

Reply:
Done locally. That was the original implementation I had before refactoring to try to maintain consistency with the rest of the code.

One thing I noticed that I missed is how to handle the edge case where 1.) both host & config-level define added TLVs, and the host-level metadata is invalid. In this case, we'd return an empty list of custom TLVs, but at a glance, I don't think this is the right behavior.

Instead, I think we'd attempt to unpack the typed metadata namespace, and if that doesn't work, log the message & avoid exiting early fallback & attempt to add any TLVs defined at the config-level to handle that error a bit more gracefully. I'll work on adding a test case for this in the next set of commits. Let me know if you think that's the wrong behavior.